File: src/main/java/com/chattriggers/ctjs/triggers/OnCommandTrigger.java
package com.chattriggers.ctjs.triggers;

import com.chattriggers.ctjs.commands.Command;
import com.chattriggers.ctjs.commands.CommandHandler;
import net.minecraftforge.client.ClientCommandHandler;

import java.util.ArrayList;
import java.util.Arrays;

public class OnCommandTrigger extends OnTrigger {
    private String commandName = null;
    private Command command = null;

    public OnCommandTrigger(Object method) {
        super(method, TriggerType.COMMAND);
    }

    @Override
    public void trigger(Object... args) {
        if (!(args instanceof String[])) throw new IllegalArgumentException("Arguments must be string array");

        callMethod(args);

    }

    /**
     * Sets the command name.<br>
     * Example:<br>
     * OnCommandTrigger.setCommandName("test")<br>
     * would result in the command being /test
     *
     * @param commandName The command name
     * @return the trigger for additional modification
     */
    public OnCommandTrigger setCommandName(String commandName) {
        this.commandName = commandName;

        reInstance();

        return this;
    }

    /**
     * Alias for {@link #setCommandName(String)}
     *
     * @param commandName The command name
     * @return the trigger for additional modification
     */
    public OnCommandTrigger setName(String commandName) {
        return setCommandName(commandName);
    }

    // helper method to re instance the command
    private void reInstance() {
        for (Command command : CommandHandler.getCommandList()) {
            if (command.getCommandName().equals(this.commandName)) {
                command.getTriggers().add(this);
                return;
            }
        }

        this.command = new Command(this, this.commandName, "/" + this.commandName);
        ClientCommandHandler.instance.registerCommand(this.command);
        CommandHandler.getCommandList().add(this.command);
    }
}

File: src/main/kotlin/com/chattriggers/ctjs/triggers/OnCommandTrigger.kt
package com.chattriggers.ctjs.triggers

import com.chattriggers.ctjs.commands.Command
import com.chattriggers.ctjs.engine.ILoader
import com.chattriggers.ctjs.utils.kotlin.External

@External
class OnCommandTrigger(method: Any, loader: ILoader) : OnTrigger(method, TriggerType.Command, loader) {
    private lateinit var commandName: String
    private val tabCompletions = mutableListOf<String>()
    private val aliases = mutableListOf<String>()
    private var command: Command? = null

    override fun trigger(args: Array<out Any?>) {
        callMethod(args)
    }

    /**
     * Sets the tab completion options for the command.
     * This method must be used before setting the command name, otherwise, the tab completions will not be set.
     *
     * @param args all the tab completion options.
     */
    fun setTabCompletions(vararg args: String) = apply {
        tabCompletions.addAll(args)
    }

    /**
     * Sets the aliases for the command.
     *
     * @param args all the aliases.
     */
    fun setAliases(vararg args: String) = apply {
        if (!::commandName.isInitialized) {
            throw IllegalStateException("Command name must be set before aliases!")
        }

        aliases.addAll(args)
        reInstance()
    }

    /**
     * Sets the command name.
     * Example:
     * OnCommandTrigger.setCommandName("test")
     * would result in the command being /test
     *
     * @param commandName The command name
     * @return the trigger for additional modification
     */
    fun setCommandName(commandName: String) = apply {
        this.commandName = commandName
        reInstance()
    }

    /**
     * Alias for [setCommandName]
     *
     * @param commandName The command name
     * @return the trigger for additional modification
     */
    fun setName(commandName: String) = setCommandName(commandName)

    private fun reInstance() {
        command?.unregister()
        command = Command(this, commandName, "/$commandName", tabCompletions, aliases)
        command!!.register()
    }
}
